keyboards/planck/rev7/matrix.c

```diff
@@ -111,13 +112,74 @@ bool matrix_scan_custom(matrix_row_t current_matrix[]) {
     return changed;
 }
 
-uint8_t encoder_quadrature_read_pin(uint8_t index, bool pad_b) {
-    pin_t pin = pad_b ? B13: B12;
-    gpio_set_pin_input_high(pin);
-    gpio_write_pin_low(matrix_row_pins[index]);
-    wait_us(10);
-    uint8_t ret = gpio_read_pin(pin) ? 1 : 0;
-    gpio_set_pin_input_low(matrix_row_pins[index]);
-    gpio_set_pin_input_low(pin);
-    return ret;
+#if defined(ENCODER_ENABLE) || defined(ENCODER_MAP_ENABLE)
+
+#if !defined(PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY)
+#   define PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY 20
+#endif
+#if !defined(ENCODER_RESOLUTION)
+#   define ENCODER_RESOLUTION 4
+#endif
+
+#define ENCODER_CLOCKWISE true
+#define ENCODER_COUNTER_CLOCKWISE false
+
+int8_t  encoder_LUT[]     = {0, -1, 1, 0, 1, 0, 0, -1, -1, 0, 0, 1, 0, 1, -1, 0};
+uint8_t encoder_state[8]  = {0};
+int8_t  encoder_pulses[8] = {0};
+
+static void encoder_handle_state_change(uint8_t index, uint8_t state) {
+    uint8_t i = index;
+
+    const uint8_t resolution = ENCODER_RESOLUTION;
+
+    encoder_pulses[i] += encoder_LUT[state & 0xF];
+
+    if (encoder_pulses[i] >= resolution) {
+        encoder_queue_event(index, ENCODER_COUNTER_CLOCKWISE);
+    }
+
+    // direction is arbitrary here, but this clockwise
+    if (encoder_pulses[i] <= -resolution) {
+        encoder_queue_event(index, ENCODER_CLOCKWISE);
+    }
+    encoder_pulses[i] %= resolution;
+
+}
+
+void encoder_driver_task(void) {
+
+    // set up C/rows for encoder read
+    for (int i = 0; i < MATRIX_ROWS; i++) {
+        gpio_set_pin_output(matrix_row_pins[i]);
+        gpio_write_pin_high(matrix_row_pins[i]);
+    }
+
+    // set up A & B for reading
+    gpio_set_pin_input_high(B12);
+    gpio_set_pin_input_high(B13);
+
+    for (int i = 0; i < MATRIX_ROWS; i++) {
+        gpio_write_pin_low(matrix_row_pins[i]);
+        wait_us(PLANCK_ENCODER_SETTLE_PIN_STATE_DELAY);
+        uint8_t new_status = (palReadPad(GPIOB, 12) << 0) | (palReadPad(GPIOB, 13) << 1);
+        if ((encoder_state[i] & 0x3) != new_status) {
+            encoder_state[i] <<= 2;
+            encoder_state[i] |= new_status;
+            encoder_handle_state_change(i, encoder_state[i]);
+        }
+        gpio_write_pin_high(matrix_row_pins[i]);
+    }
+
+    // revert A & B to matrix state
+    gpio_set_pin_input_low(B12);
+    gpio_set_pin_input_low(B13);
+
+    // revert C/rows to matrix state
+    for (int i = 0; i < MATRIX_ROWS; i++) {
+        gpio_set_pin_input_low(matrix_row_pins[i]);
+    }
+
 }
```

Comment:
This rolls back encoder abstraction PR, so I can't approve it.
Work out what the actual issue is, rather than blindly reverting the code.

Reply:
@tracyde Thank you for your review. I see. Of course, If can use abstraction codes, I think that needs use it. but does not work with abstraction code.
but I tried update code with encoder_quadrature_handle_read. what do you think? https://github.com/qmk/qmk_firmware/pull/23967/commits/1cf0cc6ed73756bd5c0e8dd7bbb6f837f0101520

Note: This message is old. @sigprof tells me good code after commit code. Thank you.

Reply:
I did update code. Please continue check. https://github.com/qmk/qmk_firmware/pull/23967/commits/e34700f2a1adc9d44a70d63e496abd4aa6aed557